item-impl/src/main/java/com/example/auction/item/impl/ItemRepository.java

```diff
@@ -0,0 +1,248 @@
+import com.example.auction.item.api.ItemStatus;
+import com.example.auction.item.api.ItemSummary;
+import com.example.auction.item.api.PaginatedSequence;
+import com.lightbend.lagom.javadsl.persistence.AggregateEventTag;
+import com.lightbend.lagom.javadsl.persistence.ReadSide;
+import com.lightbend.lagom.javadsl.persistence.ReadSideProcessor;
+import com.lightbend.lagom.javadsl.persistence.cassandra.CassandraReadSide;
+import com.lightbend.lagom.javadsl.persistence.cassandra.CassandraSession;
+import org.pcollections.PSequence;
+import org.pcollections.TreePVector;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Collectors;
+
+import static com.example.auction.item.impl.CompletionStageUtils.*;
+
+@Singleton
+public class ItemRepository {
+
+    private final CassandraSession session;
+
+    @Inject
+    public ItemRepository(CassandraSession session, ReadSide readSide) {
+        this.session = session;
+        session.underlying().thenAccept(s -> registerCodec(s, new EnumNameCodec<>(ItemStatus.class)));
+        readSide.register(PItemEventProcessor.class);
+    }
+
+    CompletionStage<PaginatedSequence<ItemSummary>> getItemsForUser(
+            UUID creatorId, ItemStatus status, int page, int pageSize) {
+        return countItemsByCreatorInStatus(creatorId, status)
+                .thenCompose(
+                        count -> {
+                            int offset = page * pageSize;
+                            int limit = (page + 1) * pageSize;
+                            CompletionStage<PSequence<ItemSummary>> items = offset > count ?
+                                    CompletableFuture.completedFuture(TreePVector.empty()) :
+                                    selectItemsByCreatorInStatus(creatorId, status, offset, limit);
+                            return items.thenApply(seq -> new PaginatedSequence<>(seq, page, pageSize, count));
+                        }
+                );
+    }
+
+    private void registerCodec(Session session, TypeCodec<?> codec) {
+        session.getCluster().getConfiguration().getCodecRegistry().register(codec);
+    }
+
+    private CompletionStage<Integer> countItemsByCreatorInStatus(UUID creatorId, ItemStatus status) {
+        return session
+                .selectOne(
+                        "SELECT COUNT(*) FROM itemSummaryByCreatorAndStatus " +
+                                "WHERE creatorId = ? AND status = ? " +
+                                "ORDER BY status ASC, itemId DESC",
+                        // ORDER BY status is required due to https://issues.apache.org/jira/browse/CASSANDRA-10271
+                        creatorId,
+                        status
+                )
+                .thenApply(row -> (int) row.get().getLong("count"));
```

Comment:
It returns `count` as a long, but only allows `LIMIT` as an int!

It's a bit of a cheat, but if there are ever more than `Integer.MAX_VALUE` results, pagination will be completely broken 😕